src/libguac/string.c

```diff
@@ -124,18 +124,30 @@ char* guac_strdup(const char* str) {
     /* Do not attempt to duplicate if the length is somehow magically so
      * obscenely large that it will not be possible to add a null terminator */
     size_t length;
-    if (guac_mem_ckd_add(&length, strlen(str), 1))
+    size_t length_to_copy = strnlen(str, n);
+    if (guac_mem_ckd_add(&length, length_to_copy, 1))
         return NULL;
 
-    /* Otherwise just copy to a new string in same manner as strdup() */
-    void* new_str = guac_mem_alloc(length);
-    if (new_str != NULL)
-        memcpy(new_str, str, length);
+    /* Otherwise just copy to a new string in same manner as strndup() */
+    char* new_str = (char*)guac_mem_alloc(length);
+    if (new_str != NULL) {
+        memcpy(new_str, str, length_to_copy);
+        new_str[length_to_copy] = '\0';
```

Comment:
This appears to change the behavior of the `guac_strdup()` function, which, previously, did not add a null terminator to the end of the string. I don't know if the previous behavior was intended or not, but it seems like this could cause some issues throughout other parts of the code that depend on the lack of a null terminator, either intentionally or not?

Also, the documentation for the function states that the provided string should already be null-terminated - maybe this should be added only in the case where the new string is shorter than the original one?

Reply:
Looks to me like `guac_strdup()` did add a NULL terminator... it added 1 to `strlen(str)` and then copied that many bytes over, which will include the terminator. What am I missing?

Also, `guac_strdup()` _should add a NULL-terminator, as it's described as behaving identically to `strdup()` (which adds a NULL-terminator), aside from the NULL check.

Reply:
> Looks to me like guac_strdup() did add a NULL terminator... it added 1 to strlen(str) and then copied that many bytes over, which will include the terminator. What am I missing?

I would agree that it copies over the NULL terminator if the original string has one, but it does not manually add one to the end of the string. But maybe I'm missing something?? Here's the current source code:

https://github.com/apache/guacamole-server/blob/3ad3b041f273ec1ed30e49be49f1adcf0edd0b61/src/libguac/string.c#L118-L137

It looks to me like it:
* Checks if `str` is `NULL`, returning `NULL` if that is the case.
* Checks the string length, returning `NULL` if it cannot fit in memory.
* Allocates the memory, and then uses `memcpy` to copy over the memory.
* Returns the string.

I don't see any explicit addition of a null terminator in that function, whereas the modified version, here, does explicitly add a null terminator, as does the `guac_strlcpy()` function earlier in the file.

Reply:
Well the `strlen()` call would certainly have undefined behavior (including probable segfault) if the input string isn't NULL-terminated. And given the documented requirement that the input must be NULL-terminated, it seems like the options are:

1) Provide invalid non-NULL-terminated input and get undefined results / possible segfault
2) Provide valid NULL-terminated input and receive a NULL-terminated result.

Reply:
Okay, so that means that adding this explicit addition of a null terminator to the new string probably doesn't matter much one way or the other - it's the "safest" option in the case that a string happens to be passed in that doesn't have a null terminator and gets through the various steps (`strlen` in particular) without causing issues.